srcCpp/Infrastructure_pro.cxx

```diff
@@ -501,21 +501,118 @@ bool configureShmemTransport(
         DDS_DomainParticipantQos& qos,
         ParameterManager *_PM)
 {
-    // Number of messages that can be buffered in the receive queue.
-    int received_message_count_max = 1024 * 1024 * 2
-            / (int) _PM->get<unsigned long long>("dataLen");
-    if (received_message_count_max < 1) {
-        received_message_count_max = 1;
-    }
+    DDS_Property_t *parentProp = DDSPropertyQosPolicyHelper::lookup_property(qos.property,
+                "dds.transport.shmem.builtin.parent.message_size_max");
+    int parent_msg_size_max = atoi(parentProp->value);
+    std::cout << "message_size_max: " << parent_msg_size_max << std::endl;
+
+    /**
+     * The maximum size of a SHMEM segment highly depends on the platform.
+     * So that, we need to find out the maximum allocable space to avoid
+     * runtime errors.
+     *
+     * Perform an incremental search of allocable space in range (minSize, maxSize).
+     */
+    RTIOsapiSharedMemorySegmentHandle handle;
+    RTI_UINT64 pid = RTIOsapiProcess_getId();
+    RTIBool success = RTI_FALSE;
+    int retcode;
+    int key = rand();
+    int minSize = parent_msg_size_max;
+    int step = 1048576 + parent_msg_size_max; // 1MB + parent_msg_size_max
+    int maxBufferSize = 60817408 + step; // 58MB
```

Comment:
Minimum size is the packet of the builtin transport for SHMEM. We need it as the minimum size for a SHMEM segment since otherwise, DDS would not even be able to send a sample.

The maximum size is there in order to have a limit on the buffer size for SHMEM so Perftest does not require unlimited resources.

Since we are doing a decremental search, we use a step of 1MB to search down a pseudo-optimal buffer size in the range [minSize, maxBufferSize].

I will add this explanation as a comment on the code.

Reply:
What is the reason behind setting the `maxBufferSize` at 58MB?

Reply:
What's happens if the `"dds.transport.shmem.builtin.parent.message_size_max"` is higher than 58MB?

Reply:
> Thanks for the explanation. What is the reason behind setting the maxBufferSize at 58MB?

58MB is used on DDS examples for FlatData and is enough to get maximum performance.

> What's happens if the "dds.transport.shmem.builtin.parent.message_size_max" is higher than 58MB?

`dds.transport.shmem.builtin.parent.message_size_max` Can it ever happen?

Reply:
I don't think so. :+1: